### PR TITLE
Remove GenASM

### PR DESCRIPTION
**Pull request by soc**

### CODE REVIEW

**Comment by soc** [comment]
```
/home/jenkins/workspace/scala-2.12.x-validate-test/build-ant-macros.xml:776: Test suite finished with 1 case failing:
fail - run/t7008-scala-defined  [compilation failed]% scalac t7008-scala-defined/ScalaClassWithCheckedExceptions_1.scala
error: Error while emitting ScalaClassWithCheckedExceptions_1.scala
assertion failed: Cannot create ClassBType from non-class symbol type E1
one error found
```

**Comment by lrytz** [comment]
great! will take a look at it begin of next week.

**Comment by soc** [comment]
The failure is right, in the sense that we would generate invalid class files otherwise:

```
class C[E1 <: Exception] {
  @throws[E1]("") def bar() {}
}
```

I think the `E1` type needs to be erased to a concrete type (the upper bound).

File: test/files/run/t7008-scala-defined.flags [hunk 1]
```diff
@@ -1 +1 @@
--Ybackend:GenASM
+
```
Comment (lrytz) [inline test/files/run/t7008-scala-defined.flags]: the flags file is empty - just delete it :)

**Comment by lrytz** [comment]
@soc i submitted a fix in https://github.com/scala/scala/pull/4820, the test of this PR passes after applying it.

**Comment by SethTisue** [comment]
We discussed this at the Scala team meeting just now, and everyone is comfortable with going ahead with the removal for M4.

**Comment by soc** [comment]
@lrytz, @SethTisue thanks, great!

File: test/files/run/t7008-scala-defined.flags (hunk 1 above)
Comment (soc) [inline test/files/run/t7008-scala-defined.flags]: Will do after I rebase this one on your fix for throws!

**Comment by soc** [comment]
Rebased!

**Comment by lrytz** [comment]
fingers crossed!

**Comment by soc** [comment]
@lrytz Did you/could you review it?

**Comment by lrytz** [comment]
sure, i will!

File: test/files/jvm/patmat_opt_no_nullcheck.check [hunk 2]
```diff
@@ -1 +1,26 @@
-bytecode identical
```
Comment (lrytz) [inline test/files/jvm/patmat_opt_no_nullcheck.check:1]: this diff output should not be in the check file. i'm not sure what's the best way to fix it - there's obviously no bug here, the only problem is that GenBCode assigns the local variable slots differently for methods `a` and `b` in https://github.com/scala/scala/blob/2.12.x/test/files/jvm/patmat_opt_no_nullcheck/Analyzed_1.scala.

I think the `sameBytecode` tests are a bit fragile anyway - we should instead just check for the actual properties (the tests file mentions "no null check"). I'm fine if you move this test to pending, I can take care of re-writing it.

**Comment by lrytz** [comment]
otherwise it looks all good to me!

**Comment by lrytz** [comment]
link to the google doc summarizing test changes: https://docs.google.com/document/d/1P1oZEL3To0mMUx98Z6PKIaFHV_80kPhase0Zbn4699Q/edit

File: test/files/jvm/patmat_opt_no_nullcheck.check (hunk 2 above)
Comment (soc) [inline test/files/jvm/patmat_opt_no_nullcheck.check:1]: Argh, sorry, this was not intentional!

If I'm not mistaken, I saw that there is another method in the ASM API, something along the lines of "semantically equal". I'll try that before moving it to pending.

Edit: No, it's in partest-extras :-)

Comment (soc) [inline test/files/jvm/patmat_opt_no_nullcheck.check:1]: Mhh, `ASMConverters.equivalentBytecode` didn't work (without digging any deeper).

I'll move it to pending for now.

**Comment by soc** [comment]
Moved the test to pending for now...

**Comment by soc** [comment]
@lrytz Is this good to go? I have trouble understanding "link to the google doc summarizing test changes:". Did you just note it, or wanted to say that I should include the link in the commit message?

**Comment by lrytz** [comment]
LGTM 🚮!

**Comment by SethTisue** [comment]
:tada:, thanks @soc! nothing feels better than deleting code

**Pull request merged**